### ISSUE TITLE
Google Map supports LatLngBounds.Builder

### ISSUE THREAD

**Issue by KingWu** [flutter/flutter#36653]
Just read the source code. Cannot find a method to build a LatLngBounds from a list of LatLng.

https://developers.google.com/android/reference/com/google/android/gms/maps/model/LatLngBounds.Builder

**Comment by adim86** [comment]
I am having the same issue, I want to build the bounds from a couple of Latitude and Longitude Coordinates but I am unable to find how to implement this in FLutter with the classes we have been given. Any ideas?

**Comment by unamed000** [comment]
same issue. We need to calculate the zoom and bounds of the map between 2 markers

**Comment by adim86** [comment]
@unamed000 I was able to solve this by adding this function and using it to calculate

`LatLngBounds boundsFromLatLngList(List<LatLng> list) {
    assert(list.isNotEmpty);
    double x0, x1, y0, y1;
    for (LatLng latLng in list) {
      if (x0 == null) {
        x0 = x1 = latLng.latitude;
        y0 = y1 = latLng.longitude;
      } else {
        if (latLng.latitude > x1) x1 = latLng.latitude;
        if (latLng.latitude < x0) x0 = latLng.latitude;
        if (latLng.longitude > y1) y1 = latLng.longitude;
        if (latLng.longitude < y0) y0 = latLng.longitude;
      }
    }
    return LatLngBounds(northeast: LatLng(x1, y1), southwest: LatLng(x0, y0));
  }
`

**Comment by unamed000** [comment]
Hi @adim86, 
So the northeast should be the maximum number of lat + lng and southwest should be minimum ?

**Comment by Nexuist** [comment]
While @adim86's solution works like a charm, I still believe it should be part of the standard API to maintain feature parity with the Google Maps JS SDK. Additionally, LatLngBounds should have a .getCenter() method like its JS counterpart does.

For me this would be useful as I'm trying to make the map show all markers at once, so I need to make a bounds that includes every marker and also dispatch a camera update that sets the target an zoom to adequate values such that the map shows every marker.

**Comment by sanisloandras** [comment]
I also need this feature, but sadly don't have the time to help in development.

**Comment by karatasemre** [comment]
adim86 thank youuuuuuuu

**Comment by DanielFortuyn** [comment]
@Nexuist Yeah that is actually somehting that I would like to achieve as well, did you manage to get something like that working by now?

**Comment by Nexuist** [comment]
@DanielFortuyn Yes, I did using some custom code mixed and matched from SO:

```dart

LatLngBounds boundsFromLatLngList(List<LatLng> list) {
  double x0, x1, y0, y1;
  for (LatLng latLng in list) {
    if (x0 == null) {
      x0 = x1 = latLng.latitude;
      y0 = y1 = latLng.longitude;
    } else {
      if (latLng.latitude > x1) x1 = latLng.latitude;
      if (latLng.latitude < x0) x0 = latLng.latitude;
      if (latLng.longitude > y1) y1 = latLng.longitude;
      if (latLng.longitude < y0) y0 = latLng.longitude;
    }
  }
  return LatLngBounds(northeast: LatLng(x1, y1), southwest: LatLng(x0, y0));
}

 void _onMapCreated(GoogleMapController controller) {
    _controller.complete(controller);
    Future.delayed(
        Duration(milliseconds: 200),
        () => controller.animateCamera(CameraUpdate.newLatLngBounds(
            boundsFromLatLngList(
                widget.locations.values.map((loc) => loc.asLatLng()).toList()),
            1)));
  }

```

**Comment by fabiomaia** [comment]
@Nexuist Thanks for the `Future.delayed` snippet, I used this to work around the fact that `newLatLngBounds` was randomly throwing for me

```
E/MethodChannel#plugins.flutter.io/google_maps_14( 8309): com.google.maps.api.android.lib6.common.apiexception.c: Error using newLatLngBounds(LatLngBounds, int): Map size can't be 0. Most likely, layout has not yet occured for the map view.  Either wait until layout has occurred or use newLatLngBounds(LatLngBounds, int, int, int) which allows you to specify the map's dimensions.
```

even though the map should be all ready in `onMapReady`.

**Comment by giorgio79** [comment]
Google team committed a getVisibleRegion method, would that be the way forward? https://github.com/flutter/flutter/issues/25755

**Comment by jose-airspace** [comment]
@Nexuist how did you get the appropriate zoom level though?

**Comment by velval** [comment]
One can use the **LatLngBounds.fromPoints(points)**  built-in method if using the flutter_map library.

**Comment by khurramsaeed** [comment]
Here is a re-written version of @Nexuist suggestion as an extension function to `LatLngBounds` class:
```dart
extension LatLngBoundsExtension on LatLngBounds {
  static fromLatLng(List<LatLng> latLngList) {
      assert(latLngList.isNotEmpty);
      double? x0, x1, y0, y1;
      for (LatLng latLng in latLngList) {
        if (x0 == null) {
          x0 = x1 = latLng.latitude;
          y0 = y1 = latLng.longitude;
        } else {
          if (latLng.latitude > x1!) x1 = latLng.latitude;
          if (latLng.latitude < x0) x0 = latLng.latitude;
          if (latLng.longitude > y1!) y1 = latLng.longitude;
          if (latLng.longitude < y0!) y0 = latLng.longitude;
        }
      }
      return LatLngBounds(northeast: LatLng(x1!, y1!), southwest: LatLng(x0!, y0!));
  }
}

**Comment by tomheno** [comment]
Null-safe version from @adim86 answer : 
```dart
LatLngBounds boundsFromLatLngList(List<LatLng> list) {
  double? x0, x1, y0, y1;
  for (LatLng latLng in list) {
    if (x0 == null || x1 == null || y0 == null || y1 == null) {
      x0 = x1 = latLng.latitude;
      y0 = y1 = latLng.longitude;
    } else {
      if (latLng.latitude > x1) x1 = latLng.latitude;
      if (latLng.latitude < x0) x0 = latLng.latitude;
      if (latLng.longitude > y1) y1 = latLng.longitude;
      if (latLng.longitude < y0) y0 = latLng.longitude;
    }
  }

  return LatLngBounds(LatLng(x1!, y1!), LatLng(x0!, y0!));
}

**Comment by Daeon97** [comment]
> @unamed000 I was able to solve this by adding this function and using it to calculate
> 
> ```
> LatLngBounds boundsFromLatLngList(List<LatLng> list) {
>     assert(list.isNotEmpty);
>     double x0, x1, y0, y1;
>     for (LatLng latLng in list) {
>       if (x0 == null) {
>         x0 = x1 = latLng.latitude;
>         y0 = y1 = latLng.longitude;
>       } else {
>         if (latLng.latitude > x1) x1 = latLng.latitude;
>         if (latLng.latitude < x0) x0 = latLng.latitude;
>         if (latLng.longitude > y1) y1 = latLng.longitude;
>         if (latLng.longitude < y0) y0 = latLng.longitude;
>       }
>     }
>     return LatLngBounds(northeast: LatLng(x1, y1), southwest: LatLng(x0, y0));
>   }
> ```

So I am trying to implement exactly this right now. Fortunately stumbled on this issue here. Thanks @adim86 for this. Works perfectly
